### ISSUE TITLE
Incorrect handling of unicode character \U00010900

### ISSUE THREAD

**Issue by a94d1945-5ce8-4c6c-a12e-04246d74ed6e** [python/cpython#89268]
BPO | [45105](https://bugs.python.org/issue45105)
--- | :---
Nosy | @ronaldoussoren, @ezio-melotti, @stevendaprano, @serhiy-storchaka, @eryksun, @maxbachmann
Files | <li>[selection.png](https://bugs.python.org/file50260/selection.png "Uploaded as image/png at 2021-09-05.13:08:09 by @stevendaprano")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-09-05.11:12:09.498>
labels = ['type-bug', '3.9', 'expert-unicode']
title = 'Incorrect handling of unicode character \\U00010900'
updated_at = <Date 2021-09-12.08:57:14.225>
user = 'https://github.com/maxbachmann'
```

bugs.python.org fields:
```python
activity = <Date 2021-09-12.08:57:14.225>
actor = 'serhiy.storchaka'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Unicode']
creation = <Date 2021-09-05.11:12:09.498>
creator = 'maxbachmann'
dependencies = []
files = ['50260']
hgrepos = []
issue_num = 45105
keywords = []
message_count = 12.0
messages = ['401078', '401079', '401082', '401083', '401084', '401086', '401088', '401090', '401092', '401095', '401656', '401658']
nosy_count = 6.0
nosy_names = ['ronaldoussoren', 'ezio.melotti', 'steven.daprano', 'serhiy.storchaka', 'eryksun', 'maxbachmann']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue45105'
versions = ['Python 3.6', 'Python 3.9']
```

</p></details>

### Linked PRs
* gh-144744

**Comment by schaiba** [comment]
Hello, is this still an issue?

**Comment by andjc** [comment]
The expected VISUAL order depends on the directionality applied to the string and bidirectional support in application you are viewing with. The string is as is, but display of the string is dependent on bidi heuristics. In the actual issue, we see `s = '0𐤀00'` but the default directionality for HTML elements is LTR. Web HTML also applies bidi isolation. For isolation you can have LTR, RTL and First Strong. What the visual results are will depend on what the overall directionality of string resolves to. 

If I explicitly wrap the string in bidi formatting characters, you can see the differences:

```py
>>> s = '0𐤀00'
>>> s.encode('utf-8')
b'0\xf0\x90\xa4\x8000'
>>> [f'{ord(char):04X}' for char in s]
['0030', '10900', '0030', '0030']
>>> LRI, RLI, FSI, PDI = ('\u2066', '\u2067', '\u2068', '\u2069')
>>> print(FSI + s + PDI)
⁨0𐤀00⁩
>>> print(LRI + s + PDI)
⁦0𐤀00⁩
>>> print(RLI + s + PDI)
⁧0𐤀00⁩
>>>
```

The Phoenician letter is at index 1, i.e. 2nd letter in the string, but its position in the visual display order depends on the directionality of the string as a whole. In many situations this is inherited as LTR

In the examples above we explicitly control directionality of the isolated directional runs.

The following begin an isolated directional run:

LRI is U+2066 LEFT-TO-RIGHT ISOLATE
RLI is U+2067 RIGHT-TO-LEFT ISOLATE
FSI is U+2068 FIRST STRONG ISOLATE

and PDI indicates the end of the isolated run, where

PDI is U+2069 POP DIRECTIONAL ISOLATE

As things stand, it is the responsibility of the developer to handle directional runs correctly at IO boundaries, and this issue is an example of the need for that care. It is necessary to distinguish between the contents of the string, and metadata like the overall direction of the string. Depending on whether environments default to first strong algorithm or LTR, or are overridden to be RTL, visual results may change. 

A useful discussion (relating to web standards) is [Strings on the Web: Language and Direction Metadata](https://www.w3.org/TR/string-meta/), I doubt there is anything similar in the python domain to string-meta. 

In summary it is not an issue with Python strings, it is an issue of how a developer handles directional runs at the IO boundary.
